File: chor-extraction/src/main/java/epp/EndPointProjection.java
package epp;

import antlr4.ChoreographyLexer;
import antlr4.ChoreographyParser;
import ast.cc.interfaces.CCNode;
import ast.sp.nodes.Network;
import ast.sp.nodes.ProcessBehaviour;
import org.antlr.v4.runtime.ANTLRInputStream;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.tree.ParseTree;

import java.util.ArrayList;
import java.util.HashSet;

public class EndPointProjection {
    private ANTLRInputStream stream;
    private ChoreographyLexer lexer;
    private ChoreographyParser parser;
    private ParseTree tree;


    private ParseTree parse(String grammar) {
        try {
            stream = new ANTLRInputStream(grammar);
            lexer = new ChoreographyLexer(stream);
            parser = new ChoreographyParser(new CommonTokenStream(lexer));
            tree = parser.program();
        } catch (Exception e){
            System.out.println(e.getMessage());
        }
        return tree;
    }

    public Network project(String choreography) throws MergingException {
        ParseTree tree = this.parse(choreography);
        ArrayList<ProcessBehaviour> network = new ArrayList<>();

        if (tree!=null) {
            ChoreographyVisitor choreographyVisitor = new ChoreographyVisitor();
            CCNode ccast = choreographyVisitor.getCCAST(tree);
            HashSet<String> processes = choreographyVisitor.getProcesses();

            BehaviourProjection behaviourProjection = new BehaviourProjection();
            for (String process: processes) {
                network.add ((ProcessBehaviour) behaviourProjection.getSPAST(ccast, process));
            }
        }
        return new Network (network);
    }
}

File: chor-extraction/src/main/java/epp/EndPointProjection.kt
package epp

import gen.ChoreographyLexer
import gen.ChoreographyParser
import ast.cc.interfaces.CCNode
import ast.sp.nodes.Network
import ast.sp.nodes.ProcessBehaviour
import org.antlr.v4.runtime.ANTLRInputStream
import org.antlr.v4.runtime.CommonTokenStream
import org.antlr.v4.runtime.tree.ParseTree
import sun.invoke.empty.Empty

import java.util.HashSet
import java.util.TreeMap

class EndPointProjection {
    private fun parse(grammar: String): ChoreographyParser.ProgramContext? {
        try {
            val stream = ANTLRInputStream(grammar)
            val lexer = ChoreographyLexer(stream)
            val parser = ChoreographyParser(CommonTokenStream(lexer!!))
            val tree = parser!!.program()
            return tree
        } catch (e: Exception) {
            println(e.message)
        }
        return null //Need to fix this
    }

    @Throws(MergingException::class)
    fun project(choreography: String): Network {
        val tree = this.parse(choreography)
        val network = TreeMap<String,ProcessBehaviour>()

        if (tree != null) {
            val choreographyVisitor = ChoreographyVisitor()
            val ccast = choreographyVisitor.getCCAST(tree)
            val processes = choreographyVisitor.processes

            val behaviourProjection = BehaviourProjection()
            for (process in processes) {
                network.put(process, behaviourProjection.getSPAST(ccast, process) as ProcessBehaviour)
            }
        }
        return Network(network)
    }
}
